### ISSUE TITLE
Discourage "it" as function parameter

### ISSUE THREAD

**Issue by forthrin** [rubocop/rubocop#14322]
## Is your feature request related to a problem? Please describe.

```ruby
def foo1(it = 'bar')
  puts it
  'baz'.then { puts it }
end

foo1

def foo2(its = 'bar')
  puts its
  'baz'.then { |its| puts its }
end

foo2
```

```
bar
bar
bar
baz
```

`Lint/ShadowingOuterLocalVariable` doesn't warn when a function parameter named `it` takes precedence over an implicit `it` in a block.

## Describe the solution you'd like

Warn about it (pun intended).

## Describe alternatives you've considered

Rather, warn about using `it` as a function parameter, which seems like asking for trouble in the first place.

## Additional context

I love `it`.

**Comment by bbatsov** [comment]
I'm guessing it'd be better to have a separate cop checking for that particular shadowing, especially given the other cop mentioned is now disabled by default. //cc @koic @dvandersluis

**Comment by forthrin** [comment]
Come to think of `it`, is the function parameter taking precedence even consciously intended behavior?

`it` seems that the opposite would be more in line with coder expectations (naively or not).

https://bugs.ruby-lang.org/issues/18980

**Comment by Earlopain** [comment]
This it not special to it being a function parameter. If you specify it as a local, you get the same behaviour. To me at least it makes sense that an explicit local takes precedence over something that is only defined implicitly.

There is `Style/ItAssignment` (added in https://github.com/rubocop/rubocop/pull/13474) but it only checks for assignments to `it` inside of a block. I think it makes sense to enhance this cop instead of adding a new one?

**Comment by dvandersluis** [comment]
There are other situations that are not covered by `Style/ItAssignment` on top of function parameters. For instance, a local var defined in the function would have the same issue:

```ruby
def foo
  it = 'bar'
  'baz'.then { puts it }
end
```

It might be useful to expand `Style/ItAssignment` to handle all `lvasgn` (not just within a block), and `arg` nodes named `it`. Even if `it` isn't called in a block currently, it could lead to future confusion. We could theoretically put this behind configuration if there's concern it'd become annoying. @bbatsov what do you think?

**Comment by dvandersluis** [comment]
I opened #14365 to handle more types of creating a local named `it`.

**Comment by forthrin** [comment]
`it` seems to work as expected. `it` catches three variants:

1. When `it` is used as a function parameter: `def is_on?(it)`
1. When `it` is explicit and superfluous: `birds.each do |it|`
1. When `it` is used with `.with_index`: `bees.with_index.each do |it, i|`

Should these be differentiated with different cops, or are they logically the same?

PS! One could envision the latter to assume `|it, i|` as default parameters if left out, eg. `bees.with_index.each { printf "Bee number #{i} (#{it}) fell in love!\n" }`. Would getting this by the Ruby maintainers be a Sisyphean undertaking?
